### ISSUE TITLE
let user sort occurence search results

### ISSUE THREAD

**Issue by gbif-portal** [gbif/portal-feedback#224]
**let user sort occurence search results**

Why don&#39;t you let users click on concepts to sort search results?

For example, sort the occurrences table of one institution by clicking on column header (taxon, country, date, family, catalognumber ...)

Another useful thing: let users sort the images gallery by taxa, so we can see together all images related to the same species or family, without needing to filter the data first (so we can easily navigate the gallery from one species to another, and QUICKLY).

-----
fbitem-90ca04c707cd953f778e8789ed7ec98e67b20053
Reported by: sant.herbarium@gmail.com
System: Chrome 49.0.2623 / Windows XP 0.0.0
Referer: https://demo.gbif.org/occurrence/gallery?publishing_org=def87a70-0837-11d9-acb2-b8a03c50a862&taxon_key=5015
Window size: width 1024 - height 707
[API log](http://elk.gbif.org:5601/app/kibana#/discover/UAT-Varnish-403s?_g=(refreshInterval:(display:Off,pause:!f,value:0),time:(from:'2017-06-27T19:40:08.623Z',mode:absolute,to:'2017-06-27T19:46:08.623Z'))&_a=(columns:!(request,response,clientip),filters:!(),index:'prod-varnish-*',interval:auto,query:(query_string:(analyze_wildcard:!t,query:'response:%3E499%20AND%20(request:%22%2F%2Fapi.gbif.org%22)')),sort:!('@timestamp',desc))&indexPattern=uat-varnish-*&type=histogram)
[Site log](http://elk.gbif.org:5601/app/kibana#/discover/UAT-Varnish-403s?_g=(refreshInterval:(display:Off,pause:!f,value:0),time:(from:'2017-06-27T19:40:08.623Z',mode:absolute,to:'2017-06-27T19:46:08.623Z'))&_a=(columns:!(request,response,clientip),filters:!(),index:'prod-varnish-*',interval:auto,query:(query_string:(analyze_wildcard:!t,query:'response:%3E399%20AND%20(request:%22%2F%2Fdemo.gbif.org%22)')),sort:!('@timestamp',desc))&indexPattern=uat-varnish-*&type=histogram)

**Comment by timrobertson100** [comment]
@fmendezh can you please comment on the feasibility of a) sorted results from SOLR and b) grouping results by e.g. species?

If I recall correctly, sorting would only work at this scale when the result set was small right?

**Comment by dgasl** [comment]
You mean results cannot be sorted by server?

In that case, perhaps you could implement some kind of client-side javascript sorting within the DIV where the images are shown.

And, if the user clicks at the bottom (to load more images from server), the div could get sorted again at client-side.

**Comment by MortenHofft** [comment]
Unless i misunderstand something then sorting client side isn't an option. 

Since the client only knows the first 25 results or so. We can of course sort those in the client. But that will only sort the first 25 results. So say sorting by dataset alphabetically will still only sort the first 25 results, not giving you everything with A or Z, but whatever (seemingly) random results that are in view. Similar for sorting the images. I believe it has to be server side if it is to have any real value. But that would be nice.

Correct me if I'm wrong and you do see a value in sorting the results known to the client.

For navigating by the gallery, we could add a way to add that taxon as a filter. So instead of - or complementary to - linking to the occurrence, we could add a filter for that species?

**Comment by dgasl** [comment]
Hi @MortenHofft 

Of course server-side sorting is THE option.
I was just answering @timrobertson100 comment. 
I absolutely agree that client-side sorting is a much worse option.  
But, if it is the only option, then it is still an option. Specially with regards to the image gallery.

You have explained it correctly. 
The **"table" tab** is paginated (20 per page), and when I click next page I lose the previous results. 
The same thing happens to **"species" tab** (about 100 per page).

But the image **"gallery" tab** does not work that way.

If my search returns many images, the server shows just the first 100 (or so).
At the bottom, there is a text letting me load "more".
If I click "more", the next 100 are added but I don't lose the previous results.
So, I was not talking about filtering the gallery by taxa.  I just meant sorting the results so I can browse them all (even if I have to click "more" 20 times to get them all).
Take in account I might not be interested in images of a particular taxon: my left-side filter could have been a collector, a country, typestatus ... whatever.   But I want to see all those images in order.

So, yes: I see a value in sorting the results known to the client (at least regarding gallery tab).
I know these are just some partial "random" results returned by server (which is bad), displaying them in no order makes them difficult to browse (which is worse).

If they are sort in a client-side-selected-order I can quickly find if they include something useful for me. Otherwise I have to check them all from start to end when I look for whatever information.
And even if the page has only 20 results, this is a hard-work.
The only "good thing" of this disorder is that final users have not a fake feeling (thinking that all results are already displayed).
But I do prefer GBIF telling me that I am seeing only 20 random results out of 75000 -so there could be more, even if the page shows data from A to Z-. A choice to sort those visible results is good (specially if I can add more results to the page).

I said "random" results because I don't know which sort concept is used by the server (I guess some kind of datelastmodified, datelastindexed, gbifID, ... but anyway, this is randomness from the final user point of view).

In summary: client-side sorting is mostly useful for the image **gallery**, since it contains a certain number of records but it is not paged.   Perhaps this choice (loading "more" results instead of loading "next" page of results) could be also available to **table** and **species** tabs?

And for sure, server-side sorting would be preferred option in case it is possible.

**Comment by MortenHofft** [comment]
See also example usage of ordering [Occurrence Search API: add first / last occurrence filter](http://dev.gbif.org/issues/browse/POR-3093?jql=project%20%3D%20POR%20AND%20status%20%3D%20Open)

Users are interested in finding out the first occurrence of a species (e.g. in a country, polygon, etc). This is important for invasive alien species investigators for example.
Similarly, it would be valuable to find out the last occurrence of a species.
The user would combine this search with filters by location and basisOfRecord according to their interest of course.
There could be one or more occurrences returned in the search result.
Alternatively, the user must sort the data themselves to draw these conclusions.
